Test suite hardening

Dridi opened the pull request:
This is work in progress, following #207 I already pushed some improvements but I feel that these changes should run on more than just my system.

Don't hesitate to start reviewing it sooner than later, even though it is far from complete. This should give a precise idea of the direction I'm taking, and how writing test cases should be overall more readable and less error-prone if it proves to be a good direction.

Dridi commented:
One of my concerns regarding other platforms is the need of `lsof` that may not have everything we need on older platforms. Also `lsof` detection isn't done yet in the configure script.

Dridi commented:
This is almost complete, the remaining bits from the old setup that I need to go over is now tiny:

```
# begin old setup

export LC_ALL=C

LISTENPORT=`expr $$ % 62000 + 1024`
CERTSDIR="${TESTDIR}/certs"
CONFDIR="${TESTDIR}/configs"

# end old setup
```

The only part for which I don't have a clear plan is ensuring that no port collisions may occur. I will likely add a new test helper for ports but keep today's best-effort poor-man's pseudo-pseudo-random port selection.

Then I'll document the helpers a bit, and see how to further improve the readability of test reports.

See XXX comments in the test suite to get an idea of where polish is needed.

daghf commented:
I've only had a quick look for now, but this is looking really good!

I think the following is the only thing needed to fix make dist, with that I'd be happy to merge.

```
diff --git a/src/Makefile.am b/src/Makefile.am
index 40500e0..e72a86e 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -3,7 +3,7 @@ sbin_PROGRAMS = hitch
 noinst_LIBRARIES = libcfg.a libforeign.a
 
 EXTRA_DIST = \
-       $(top_srcdir)/src/tests/common.sh \
+       $(top_srcdir)/src/tests/hitch_test.sh \
        $(top_srcdir)/src/tests/test*.sh \
        $(top_srcdir)/src/tests/certs/* \
        $(top_srcdir)/src/tests/configs/default.cfg
```

Dridi commented:
This is embarrassing...

However before fixing this, can you have a look at the XXX comments I added to the test suite? At least those present in test cases:

```
$ git grep XXX src/tests/test*.sh
src/tests/test07-nomatch-abort.sh:# XXX: why don't we expect 'unrecognized name' again?
src/tests/test08-test-configs.sh:# XXX: unclear check
src/tests/test11-cfg.sh:# XXX: reload only works with absolute paths
src/tests/test15-proxy-v2-npn.sh:# XXX: why do we expect ALPN and not NPN here?
src/tests/test23-tls-protos-tls1_2.sh:# XXX: find how to detect the lack of SSLv3
src/tests/test24-tls-protos-tls1_1.sh:# XXX: find how to detect the lack of SSLv3
```

Since this work is incomplete, I may push the rest directly if you agree.

Dridi commented:
Rebased against master, but not ready to merge because the test cases introduced since this pull request was opened aren't migrated yet.

When hitch fails, we are currently blind in the test report. For that I added e4a115effb44b14905315ddefff7b47479c995df to dump any `*.log` files found upon failure. Currently that means adding manually `log-filename = "something.log"` when `--config` is used. Having the ability to specify a log file on the command line could be used transparently by all tests. A `--verbose` option would be nice too to "unquiet" the `--daemon` option for test cases.

klenan commented:
OpenBSD doesn't have `losof(8)`. Getting rid of in in the test suite is still on my todo but I'll be happy if anyone beats me to it.

Dridi commented:
@klenan, thanks for letting us know, I can take care of it and switch to `sockstat(1)` instead. This pull request is waiting until master is in a merge-able state anyway so I'm not pushing anything until other known issues are solved and remaining tests are ported.

Dridi commented:
Rebased against current master, with the addition of 8d796e591db5c06e228f2568b61894af84fe8488 for the tests that were added after the pull request was started. Previous test failures are gone, but test 27 shows relative paths problems.

I believe we can merge this into master, and keep the remaining enhancements for later. One more time, please pay attention to the XXX comments throughout the test suite.

klenan commented on src/tests/test23-tls-protos-tls1_2.sh:
```diff
@@ -1,23 +1,21 @@
 # this will fail on platforms that have OpenSSL compiled without SSLv3
-openssl s_client -connect $LISTENADDR:$LISTENPORT -tls1_2
-test $? -eq 0 || die "Connecting using TLS 1.2 failed."
+# XXX: find how to detect the lack of SSLv3
```
`s_client -ssl3 | grep -Fq 'unknown option -ssl3'`?

klenan commented on src/tests/test08-test-configs.sh:
```diff
@@ -98,23 +104,16 @@ daemon = "on"
-test $? -eq 0 || die "--help after --config does not work as expected."
+run_cmd hitch --config=${CONFDIR}/default.cfg --help
 
+# XXX: unclear check
```
See the referenced issue #52 (and #22).

You seem to have removed the successfull return check after the `--config=... --help` invocation, I suggest adding that back in and also leaving the `--test --config=...` test in place.

Dridi replied:
I'm confused by the error message not matching the actual command (`--test` vs `--help`) but with #52 and #22 I can see it was probably a matter of copy/pasta. I will re-enable this check, thanks.

klenan commented on src/tests/test23-tls-protos-tls1_2.sh:
```diff
@@ -1,23 +1,21 @@
 # this will fail on platforms that have OpenSSL compiled without SSLv3
-openssl s_client -connect $LISTENADDR:$LISTENPORT -tls1_2
-test $? -eq 0 || die "Connecting using TLS 1.2 failed."
+# XXX: find how to detect the lack of SSLv3
```
One naive way would be to check for the `-ssl3` flag missing (note how `-no_ssl3` is still there):
`$ openssl s_client -help 2>&1 | grep -Fq -- -ssl3`

Dridi replied:
We don't want to run `s_client` in a sub-shell (referring to a previous comment) and `-ssl3` wasn't used in the the original test, so I'd need to build `openssl`and find out myself. Otherwise the new check is as comprehensive as the old one. I should probably document how to use the test helpers, I was at least planning to...

But we could do something like this:

```
openssl s_client -help 2>&1 |
grep -q -e -ssl3 ||
skip "no ssl3 support"
```

Same as here:

https://github.com/Dridi/hitch/blob/8d796e591db5c06e228f2568b61894af84fe8488/src/tests/test25-dynamic-backend-address.sh#L10-L12

Dridi commented:
Ready to be merged IMO, and I will retire `lsof` with a new patch later. The last commit in the list documents the new shell functions and how they are meant to be used. I encourage anyone dealing with the test suite to have a look.

Cheers!

daghf commented:
No complaints from me - great work.

With any remaining issues(?) it's regardless a big net improvement from the current testing setup, so please go ahead and merge.

Dridi commented:
Instead of merging I decided to get a quick stab at `lsof` independence, and added 6f8cc44, because my last `make check` before hitting the merge button didn't go well.

After upgrading to Fedora 27 I have one test failure and one test error (consistently). Off the top of my head one possible related change is that now Fedora is building `curl` with openssl instead of GnuTLS (but I'm not certain this change was made, poor memory).

> [ 1299] {core} Child 1300 was terminated by signal 11.

In order to implement the `sockstat` support I took 10 minutes to spin up a FreeBSD 11 system, and while the patch was trivial to write, I'm still left performing the ancient ritual of head-scratching trying to understand why two tests fail and two tests err.

> [12026] WARNING: {core} Unable to notify worker 12045 with changed backend address (Bad file descriptor).

@klenan, please let us know how things pan out on OpenBSD with the latest patch.

Courtesy of my recent OS upgrade ;)

daghf commented:
I think the test errors you are seeing are due to breakage in the backend refresh bits, hopefully addressed by @dmatetelki in PR #224

dmatetelki commented:
@dridi: My PR #224 fixes some broken things, including the "child terminated by signal 11". I would vote for get this PR in, so I can apply mine.

Dridi commented:
@klenan, I'm getting this increasing pressure to merge from the team. Once you have time to test this, please either comment here if it works fine on OpenBSD or open a new issue if it doesn't.

Thanks for the reviews!
